doc/spec.md

```diff
@@ -2348,7 +2355,62 @@ foo({ id: 1234, name: false });    // Error, name of wrong type
 foo({ name: "hello" });            // Error, id required but missing
 ```
 
-### <a name="3.11.5"/>3.11.5 Excess Properties
+### <a name="3.11.5"/>3.11.5 Comparability
+
+*TODO: Document the base primitive type.
+
+*Types are required to be *comparable* in certain circumstances, such as part of when checking whether two values of given types might be equal at runtime using operators like '===', or when using a type assertion.
+
+*S* is ***comparable to*** a type *T*, and *T* is ***comparable from*** *S*, if one of the following is true:
+
+* *S* and *T* are identical types.
+* *S* or *T* is the Any type.
+* *S* or *T* is an enum type and the other is the primitive type Number.
+* *S* is a literal type and *T* is the base primitive type of S.
+* *S* is a union type and some constituent type of *S* is comparable to *T*.
+* *S* is an intersection type and at least one constituent type of *S* is comparable to *T*.
+* *T* is a union type and *S* is comparable to at least one constituent type of *T*.
+* *T* is an intersection type and *S* is comparable to each constituent type of *T*.
+* *S* is a type parameter and the constraint of *S* is comparable to *T*.
```

Comment:
Above this line:

* *T* is a type parameter and *S* is mapped type `{ [P in keyof T]: X }` and *X* is related to *T[P]*.
* *T* is a keyof type `keyof T'` and
  * *S* is a keyof type `keyof S'` and *T'* is comparable to *S'*, or
  * *S* is related to `keyof C`, where *C* is the constraint of *T'*.
* *T* is an indexed access type `T'[K]` and *S* is related to *C[K]*, where *C* is the constraint of *T'*.
* *S* is a mapped type `{ [P in keyof S]: X }` and *S[P]* is related to *X*.

Reply:
On this line, append "and *T* is not the `object` type."

Reply:
Below this line:

* *S* is an indexed access type `S'[K]` and *C[K]` is related to *T*, where *C* is the constraint of *S'*.

Reply:
I'm not sure what some of these corrections refer to, but should

* *S* is a mapped type `{ [P in keyof S]: X }` and *S[P]* is related to *X*.

be

* *T* is a mapped type *{ [P in K]: X }*, *S* is not a mapped type, *keyof S* is identical to *K*, and *S[K]* is comparable to *X*.

Reply:
Yes, looks like I just messed that one up. It should be "*S[P]* is comparable to *X*", though. You could write it more compactly as:

* *T* is a mapped type *{ [P in keyof S]: X }* and *S[P]* is comparable to *X*, where *S* is not a mapped type.

Since you can express the identity of *K* and *keyof S* implicitly.